### ISSUE TITLE
Problems with optimizing X25519

### ISSUE THREAD

**Issue by mkannwischer** [slothy-optimizer/slothy#38]
I was trying to put together an example showing how to use the splitting heuristic on the X25519. 
I'm not trying to have optimal code (for that we of course have https://github.com/slothy-optimizer/slothy/blob/main/paper/scripts/slothy_x25519.sh), but just trying to have something that terminates in a reasonable time on a laptop and demonstrates how the splitting heuristic works. 

I have the following code (to be placed in a Python script in the root dir):
```
import logging
import sys

from slothy import Slothy

import slothy.targets.aarch64.aarch64_neon as AArch64_Neon
import slothy.targets.aarch64.cortex_a55 as Target_CortexA55

logging.basicConfig(stream=sys.stdout, level=logging.INFO)

arch = AArch64_Neon
target = Target_CortexA55

slothy = Slothy(arch, target)

# example
slothy.load_source_from_file("examples/naive/aarch64/X25519-AArch64-simple.s")

slothy.config.inputs_are_outputs=True
slothy.config.outputs=["x0"]
slothy.config.variable_size=True
slothy.config.constraints.stalls_first_attempt=32
slothy.config.split_heuristic = True
slothy.config.split_heuristic_stepsize = 0.1
slothy.config.split_heuristic_factor = 10
slothy.config.split_heuristic_repeat = 2
slothy.optimize(start="mainloop", end="end_label")
slothy.write_source_to_file("test.s")
```
Running this very quickly errors with
```
INFO:slothy:Instructions in body: 958
INFO:slothy.mainloop.split.0_96:Perform internal binary search for minimal number of stalls...
INFO:slothy.mainloop.split.0_96:Attempt optimization with max 32 stalls...
INFO:slothy.mainloop.split.0_96:Objective: minimize number of stalls
INFO:slothy.mainloop.split.0_96:Invoking external constraint solver (OR-Tools CP-SAT v9.7.2996) ...
INFO:slothy.mainloop.split.0_96:[4.5895s]: Found 1 solutions so far... objective 12.0, bound 6.0 (minimize number of stalls)
INFO:slothy.mainloop.split.0_96:OPTIMAL, wall time: 6.632225 s
INFO:slothy.mainloop.split.0_96:Booleans in result: 8991
INFO:slothy.mainloop.split.0_96.selfcheck:OK!
INFO:slothy.mainloop.split.0_96:Minimum number of stalls: 12
INFO:slothy.mainloop.split.96_192:Perform internal binary search for minimal number of stalls...
INFO:slothy.mainloop.split.96_192:Attempt optimization with max 32 stalls...
ERROR:slothy.mainloop.split.96_192:Ran out of registers trying to _statically_ assign architectural registers
                for input and outputs of the snippet. You should consider enabling register
                renaming for input and output, by setting config.rename_{inputs,outputs}.
                Alternatively, you may fix input and output registers by hand yourself.
Traceback (most recent call last):
  File "/home/mjk/git/slothy-tutorial/tst.py", line 34, in <module>
    slothy.optimize(start="mainloop", end="end_label")
  File "/home/mjk/git/slothy-tutorial/slothy/core/slothy.py", line 249, in optimize
    early, core, late, num_exceptional = Heuristics.periodic(body, logger, c)
                                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/mjk/git/slothy-tutorial/slothy/core/heuristics.py", line 310, in periodic
    res = Heuristics.linear( body, logger=logger, conf=conf)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/mjk/git/slothy-tutorial/slothy/core/heuristics.py", line 387, in linear
    return Heuristics._split(body, logger, conf)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/mjk/git/slothy-tutorial/slothy/core/heuristics.py", line 782, in _split
    return Heuristics._split_inner(body, logger, c)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/mjk/git/slothy-tutorial/slothy/core/heuristics.py", line 723, in _split_inner
    cur_body, stalls, _ = optimize_chunks_many(idx_lst, cur_body, stalls,show_stalls=False)
                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/mjk/git/slothy-tutorial/slothy/core/heuristics.py", line 683, in optimize_chunks_many
    body, stalls, cur_stalls, local_perm = optimize_chunk(start_idx, end_idx, body,
                                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/mjk/git/slothy-tutorial/slothy/core/heuristics.py", line 659, in optimize_chunk
    result = Heuristics.optimize_binsearch(cur_body,
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/mjk/git/slothy-tutorial/slothy/core/heuristics.py", line 132, in optimize_binsearch
    return Heuristics.optimize_binsearch_internal(source, logger, conf, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/mjk/git/slothy-tutorial/slothy/core/heuristics.py", line 234, in optimize_binsearch_internal
    success = core.optimize(source, **kwargs)
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/mjk/git/slothy-tutorial/slothy/core/core.py", line 1040, in optimize
    self._load_source(source, prefix_len=prefix_len, suffix_len=suffix_len)
  File "/home/mjk/git/slothy-tutorial/slothy/core/core.py", line 1164, in _load_source
    self._restrict_input_output_renaming()
  File "/home/mjk/git/slothy-tutorial/slothy/core/core.py", line 1288, in _restrict_input_output_renaming
    raise e
  File "/home/mjk/git/slothy-tutorial/slothy/core/core.py", line 1281, in _restrict_input_output_renaming
    v.reg = get_fresh_renaming_reg(v.ty)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/mjk/git/slothy-tutorial/slothy/core/core.py", line 1273, in get_fresh_renaming_reg
    raise OutOfRegisters
slothy.core.core.SlothyBase._restrict_input_output_renaming.<locals>.OutOfRegisters
```

If I instead do two calls to `optimize` to first resolve the symbolic variable names (in the way the other script does it), it works fine: 
```
import logging
import sys

from slothy import Slothy

import slothy.targets.aarch64.aarch64_neon as AArch64_Neon
import slothy.targets.aarch64.cortex_a55 as Target_CortexA55

logging.basicConfig(stream=sys.stdout, level=logging.INFO)

arch = AArch64_Neon
target = Target_CortexA55

slothy = Slothy(arch, target)

# example
slothy.load_source_from_file("examples/naive/aarch64/X25519-AArch64-simple.s")

slothy.config.inputs_are_outputs=True
slothy.config.outputs=["x0"]
slothy.config.constraints.functional_only = True
slothy.config.constraints.allow_reordering = False
slothy.optimize(start="mainloop", end="end_label")


slothy.config.constraints.functional_only = False
slothy.config.constraints.allow_reordering = True
slothy.config.variable_size=True
slothy.config.constraints.stalls_first_attempt=32
slothy.config.split_heuristic = True
slothy.config.split_heuristic_stepsize = 0.1
slothy.config.split_heuristic_factor = 10
slothy.config.split_heuristic_repeat = 2
slothy.optimize(start="mainloop", end="end_label")
slothy.write_source_to_file("test.s")
```

I'm not sure if this is a bug, or if I misunderstand how SLOTHY handles symbolic register names. 
@hanno-becker, any ideas?

**Comment by hanno-becker** [comment]
@mkannwischer As far as I understand, this is not a bug, but a legitimate issue one can face when working with symbolic registers: When SLOTHY attempts to optimize code which has symbolic inputs and/or outputs, it will, by default, try to _statically_ assign architectural registers to those inputs/outputs. If there are simply not enough such registers for a static assignment, it will fail. In this case, as the error message indicates, one can try letting the solver pick the assignment for symbolic input/output registers. You can try doing this here by setting `config.rename_{inputs,outputs}`.

To make it concrete, here's an example that a static assignment (which does not look at the code _at all_) can only handle with 8 registers, even though it _could_ be handled with 4 (by overlapping input and output):
```
ldr input0, [...]
ldr input1, [...]
ldr input2, [...]
ldr input3, [...]
add output0, input0, input0
add output1, input1, input1
add output2, input2, input2
add output3, input3, input3
```

_However_, even if `rename_{input, output}` gets you out of trouble in every individual step of the splitting heuristic, overall it would still not work with symbolic registers: The issue is that if you rename registers at the 'boundary' (input/output) of the current splitting window, the rest of the code somehow needs adjusting! But it could well happen that the surrounding code uses the registers that you have chosen to accommodate the symbolic input/outputs of the current window, and then you have conflicts.

I did try out some ways around this at some point -- unfortunately, I don't recall the details -- but came to the conclusion that trying to use the splitting heuristic with register renaming at the window boundaries just doesn't work. One thing one _could_ try is relax SLOTHY so that it just doesn't try to resolve symbolic registers but keeps them symbolic -- I think this would be a useful thing -- but that's not yet implemented.

TL;DR: You can't use the splitting heuristic for code using symbolics. If you want to, resolve symbolic globally first, as done in `slothy_x25519.sh`.

**Comment by mkannwischer** [comment]
Thanks! Understood.
